src/HDF5.jl

```diff
@@ -426,10 +426,11 @@ hash(x::HDF5ReferenceObj, h::UInt) = hash(x.r, h)
 # Compound types
 # These are "raw" and not mapped to any Julia type
 type HDF5Compound
-    data::Array{UInt8}
+    data::Array
```

Comment:
This should be `Array{Any}` to match your intention of storing appropriately typed data in here.